touchforms/formplayer/static/formplayer/script/entrycontrols_full.js

```diff
@@ -1,785 +1,584 @@
-  }
 }
+PhoneEntry.prototype = Object.create(IntEntry.prototype);
+PhoneEntry.prototype.constructor = IntEntry;
 
-function FreeTextEntry (args) {
-  inherit(this, new SimpleEntry());
-
-  args = args || {};
-  this.domain = args.domain || 'full';
-  this.length_limit = args.length_limit || 10000;
-  this.textarea = args.prose;
-
-  this.inputfield = null;
-  this.default_answer = null;
-
-  this.load = function (q, $container) {
-    this.mkWidget(q, $container);
-
-    this.setAnswer(this.default_answer);
-  }
-
-  this.mkWidget = function (q, $container) {
-    if (!this.textarea) {
-      $container.html('<textarea id="textfield" maxlength="' + this.length_limit + '" type="text" style="width: ' + this.widgetWidth() + '; height: 1.2em" /><span id="type" style="margin-left: 15px; font-size: x-small; font-style: italic; color: grey;">(' + this.domainText() + ')</span>');
-      var widget = $container.find('#textfield');
-    } else {
-      $container.html('<textarea id="textarea" style="width: 33em; height: 10em; font-family: sans-serif;"></textarea>');
-      var widget = $container.find('#textarea');
-    }
-    this.inputfield = widget[0];
-    widget.change(function() { q.onchange(); });
-  }
-
-  this.getControl = function () {
-    return this.inputfield;
-  }
-
-  this.getRaw = function () {
-    var control = this.getControl();
-    return (control != null ? control.value : null);
-  }
-
-  this.getAnswer = function () {
-    var raw = $.trim(this.getRaw());
-    return (raw == '' ? null : raw);
-  }
-
-  this.setAnswer = function (answer, postLoad) {
-    var control = this.getControl();
-    if (control) {
-      control.value = (answer != null ? answer : '');
-    } else {
-      this.default_answer = answer;
-    }
-  }
-
-  this.prevalidate = function (q) {
-    var raw = this.getRaw();
-    if (raw) {
-      var errmsg = this._prevalidate(raw);
-      if (errmsg) {
-        q.showError(errmsg);
-        return false;
-      }
-    }
-    return true;
-  }
 
-  this._prevalidate = function (raw) {
-    return null;
-  }
+/**
+ * The entry that defines an float input. Only accepts real numbers
+ */
+function FloatEntry(question, options) {
+    FreeTextEntry.call(this, question, options);
+    this.templateType = 'float';
 
-  this.domainText = function() {
-    return 'free-text';
-  }
+    this._prevalidate = function(raw) {
+        return (isNaN(+raw) ? "Not a valid number" : null);
+    }
 
-  this.widgetWidth = function() {
-    return '20em';
-  }
+    this.domainText = function() {
+        return 'decimal';
+    }
 }
+FloatEntry.prototype = Object.create(IntEntry.prototype);
+FloatEntry.prototype.constructor = IntEntry;
+
+/**
+ * Represents a checked box entry.
+ */
+function MultiSelectEntry(question, options) {
+    var self = this;
+    Entry.call(this, question, options);
+    self.templateType = 'select';
+    self.choices = question.choices;
+    self.choicevals = question.choicevals ? question.choicevals() : null;
```

Comment:
confused about when this was ever used. the api doesn't return the choice values (only the labels) so we use the index to answer the question. if that's the case i don't think this is necessary